### ISSUE TITLE
pty_process issues with Python 3.7+

### ISSUE THREAD

**Issue by jcohen02** [radical-cybertools/radical.saga#752]
I wasn't sure whether _Python 3.7+_ is intended to be supported at present or whether this is still WIP (the classifiers in [setup.py](https://github.com/radical-cybertools/radical.saga/blob/devel/setup.py) only list _3.5_) but I'm experiencing an issue creating a PTY in Python 3.7/3.8. This is not present in 3.5/3.6.

The issue should be reproducible using the following example code:

```
from radical.saga.utils import pty_process as pp
from radical.saga.utils import pty_shell_factory as psf
from radical.utils.logger import Logger

cfg = {'ssh_share_mode': 'auto', 'ssh_copy_mode': 'sftp', 'ssh_timeout': '10.0', 'shell': None, 'connection_pool_size': '10', 'connection_pool_wait': '600', 'prompt_pattern': '[\\$#%>\\]]\\s*$', 'connection_pool_ttl': '600'}
pty = pp.PTYProcess("/usr/bin/env TERM=vt100  \"/bin/tcsh\"  -i", cfg)
pty.alive()

f = psf.PTYShellFactory()

log = Logger('test')

info = {}
info['posix'] = True
info['key_pass'] = {}
info['host_str'] = 'localhost'
info['pass'] = ''
info['prompt'] = '[\$#%>\]]\s*$'
info['latency'] = 0.25
info['ssh_timeout'] = 10.0
info['logger'] = log

f._initialize_pty(pty, info)

pty.write('ls\n')
print(pty.read())
print(pty.read())
print(pty.read())
```

In 3.5/3.6 I get the output of `ls` printed to the screen.

In 3.7/3.8, the script hangs for some time and then fails saying that it `Could not detect shell prompt (timeout)`:

```
    f._initialize_pty(pty, info)
  File "/.../radical.saga/src/radical/saga/utils/pty_shell_factory.py", line 420, in _initialize_pty
    raise ptye.translate_exception (e)
  File "/.../radical.saga/src/radical/saga/utils/pty_shell_factory.py", line 293, in _initialize_pty
    raise rse.NoSuccess("Could not detect shell prompt (timeout)")
radical.saga.exceptions.NoSuccess: Could not detect shell prompt (timeout) (/.../radical.saga/src/radical/saga/utils/pty_shell_factory.py +293 (_initialize_pty)  :  raise rse.NoSuccess("Could not detect shell prompt (timeout)"))
```

I've spent quite some time trying to debug this and it looks like it could be locking related.  The cause of the issue looks to be that when `select.select` is called in [`pty_process.py#L593`](https://github.com/radical-cybertools/radical.saga/blob/57bc8dd09601333ffe29cc6b124cddb58967639e/src/radical/saga/utils/pty_process.py#L593) to check if there's any data to read, `rlist` is an empty list. In 3.5/3.6, it's correctly returning the file descriptor.

I tried some standalone test code using only the core Python library and this seems to work. However, when the code is run within a `radical.utils.RLock` context manager, as it is in the main `radical.saga` code, the problem occurs.

(EDIT: This issue relates to the `devel` branch)

**Comment by andre-merzky** [comment]
Hi Jeremy, 

apologies that this took again a long time to reply :-(  

I tried to reproduce the issue with the current release on Python 3.7.3, and seem to get the expected result.  Would you be so kind and verify that the problem still exists for you?

Best, Andre.

**Comment by jcohen02** [comment]
Hi Andre, thanks for investigating. I can confirm that this does now seem to be resolved when at the latest head of the `devel` branch in both `radical.saga` and `radical.utils`.

I guess this is not of critical importance at the moment but if you do want to try and identify the cause, it looks like this was an issue in `radical.utils` and it has been fixed somewhere between `v0.90.3` and `v1.0.0`.

If I put `radical.saga` back to the commit I was originally at when I had this issue (which shows up as v0.90.0 when installed - 57bc8dd09601333ffe29cc6b124cddb58967639e), I can then switch `radical.utils` between tag `v0.90.3` where the problem occurs and `v1.0.0` where it doesn't.

**Comment by andre-merzky** [comment]
Yes, indeed - RU introduced some changes to the config system during those releases, and RS was slow to catch up with those changes...

We'll push out a new release later today or tomorrow, so what you now see in devel should be the default again.

Thanks for your patience with us! :-)
